### ISSUE TITLE
2015.8.0 winrepo downloader corrupts some installers

### ISSUE THREAD

**Issue by TheBigBear** [saltstack/salt#27093]
@UtahDave @twangboy 

Not sure when exactly it started happening, but the new winrepo 2015.8.0 downloader code now corrupts some installers as it downloads them, or maybe it is the download sites that have introduced redirects, it is not necessarily something that has changed on the salt side.

I have checked the md5sum of the same installer downloaded using the URL from the winrepo sls file by a browser and compared it to the file in the relevant local ```C:\salt\var\cache\salt\minion\extrn_files\base\```  cache and they differ from each other.

the winrepo sls files used to have 
```
http://download.macromedia.com/get/flashplayer/current/licensing/win/install_flash_player_18_plugin.msi
and
http://download.macromedia.com/get/flashplayer/current/licensing/win/install_flash_player_18_active_x.msi
```

as their ```installer:``` s set. 

But those URLs have 302 re-directs to the actual download URLs which is what is now in the winrepo sls file, as of yesterday. 

```
http://fpdownload.macromedia.com/get/flashplayer/current/licensing/win/install_flash_player_18_plugin.msi
and
http://fpdownload.macromedia.com/get/flashplayer/current/licensing/win/install_flash_player_18_active_x.msi
```

After I discovered that, the file downloaded by the winrepo downloader and the files downloaded from any broswer (tested in safari, firefox, IE and chrome), I opened up the installer file with an editor and found html redirect code in the first few lines of the msi installer

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>302 Found</title>
</head><body>
<h1>Found</h1>
<p>The document has moved <a href="http://fpdownload.macromedia.com/get/flashplayer/current/licensing/win/install_flash_player_18_plugin.msi">here</a>.</p>
</body></html>
ÐÏà¡±á                >  þÿ	            1            8      €6         €                €  ÿ  €  ÿ                    	  	  ÿ	  €
```

You can easily reproduce this by trying to install 
adobeflashplayerplugin
adobeflashplayeractivex
using the winrepo files from before yesterday (13 Sept 2015). 

on the salt master all you get is:
```
    adobeflashplayeractivex:
        ----------
        failed:
            This installation package could not be opened.  Contact the application vendor to verify that this is a valid Windows Installer package.
```

I think the sooner we implement checksums/hashes for the installers in the sls file metadata, the better. As this should catch this sort of curruption and report the error accordingly. (any of the unix/linux pkg managers already have checksums built-in yum,apt,rpm, etc ...)
But in the case of winrepo I think the salt project has to do the checksumming as it already does in having to provide the winrepo sls file in the first place.

**Comment by TheBigBear** [comment]
Is there an easy way to invoke the salt python winrepo downloader, pass it an URL and have it download the file to current dir? I would quite like to compare the checksums of the files it downloads with what wget or curl would get. To see how many other installers are affected in this way?

**Comment by TheBigBear** [comment]
I have changed the URL to the redirected one. That solves the issue for these.

**Comment by TheBigBear** [comment]
Sorry, this did NOT fix it, as promising as it looked. So I am afraid this one has to be re-opened. ;-(

**Comment by TheBigBear** [comment]
Changing the URL did fix it after all. I had more than one git repos (left-over from 2015.8.0rc testing and that one 'won' and it kept trying to install the old URL.

**Comment by TheBigBear** [comment]
OK, so changing the URL in this case has solved the issue at hand. But it can still happen at any time and can be triggered by an outside side hositng an installer simply using a temporary or permanent web redirect. 

I suggest two steps should be taken.

1. examine why the salt winrepo downloader fails to properly follow and apply the web redirect?
If I understand the code correct, it should be the http.query that should handle this part more correctly?


2. we should introduce some sort of checksuming of the installers as part of the winrepo sls metadata collected (allowing for any or multiples of: crc32, md5, sha1, sha2, sha224, sha256, sha384, sha512) as requested in some other PRs https://github.com/saltstack/salt/issues/14872

**Comment by twangboy** [comment]
@TheBigBear This seems to be a problem with the salt.utils.http.query function being called by the fileclient. It's dumping everything that's returned to the file. @terminalmage is looking into it right now.

**Comment by terminalmage** [comment]
This looks to be an issue with the tornado library we are using starting with 2015.8.0. I've filed an upstream bug: https://github.com/tornadoweb/tornado/issues/1518

**Comment by TheBigBear** [comment]
OK, thanks for looking at this. This is quite qide spread and a big number of winrepo files no longer download correctly. 

Another large category is any installers that are being hosted on amazon's S3, as the 'normal' way (the only way?) to get to those files is through a very elaborate web redirect (302) that adds api keys and pins and cookies and all.

one of quite a few of those is the ```git``` (the new git ver. 2.5 - NOT the 'old' msysgit 1.x)

a nice way to look at the headers or a URL is to use curl.
```curl -D myheaders https://my-installer-on-s3 > /dev/null```
and then examine the ```myheaders``` file.

**Comment by terminalmage** [comment]
@TheBigBear I have a workaround for this which I will be submitting later tonight. Basically, we check the size of the file that was downloaded and if it's greater than the ``Content-Length`` header from the 200 OK HTTP response, then we backup the file we just downloaded and open it up for reading, skip to the start of the actual file, and start reading chunks of it and writing a new, uncorrupted file.

**Comment by twangboy** [comment]
Yuck....

**Comment by TheBigBear** [comment]
@terminalmage great news. 
The 2015.8 winrepo is basically un-useable right now as a high percentage of it's files are corrupted during download. I am very close to having to revert back, so I am looking forward to this, albeit a bit 'yucki', work-around.
Thanks.

**Comment by cachedout** [comment]
@TheBigBear I presume that we're now ready to close this since #27163 is merged?

**Comment by TheBigBear** [comment]
@cachedout yes, just finished buidling my own new windows salt minion installer and re-testing this. Works as expected now, and has the appropriate warnings in the log. Great work. Thanks.

**Comment by TheBigBear** [comment]
@twangboy this is a __signicficant__ fix for the windows salt-minion on 2015.8.0, without it they cannot install very many software pkgs at all. 

I would suggest that a new 2015.8 branch windows salt-minion 2015.8.0-4 is cut/built and put on repo.saltstack.com/windows ?

I have to add that for my own tests I have also built my windows salt-minion with python ver. 2.7.10 and this also takes care of the SSL server check/failure  https://github.com/saltstack/salt/issues/27081  that we get on all sites hosting installers that require SNI (server name indication) that python 2.7.8 does not yet support. I realise that there are other tests that I don't even know about that might well be required before such a fundamental python version upgrade. But jst thought I'll mention it here and promise to give you feedback the moment I notice other things breaking because of the newer python version 2.7.10. This SSL error is given EVEN for non secure simple http URLs.

**Comment by twangboy** [comment]
@TheBigBear I'm working on #25440 today. Python 2.7.10 may have broken other functions of the salt-minion on Windows. I can't remember what when I last tested. I'll try it again with 2015.8 and see what the issues are.
